### PR TITLE
aligned | characters

### PR DESCRIPTION
**Pull request by vekemans**
Alignment of "|" characters in the output.

### CODE REVIEW

**Comment by josharian** [comment]
Thanks! Should the alignment take into account the `(inserted)` and `(deleted)` annotations?

I wonder whether this would be simpler using `text/tabwriter`. What do you think?

**Comment by vekemans** [comment]
I wouldn't take ```(inserted)``` and ```(deleted)``` into account since this lines will probably be the longest and might look weird. Or maybe we could add the ```inserted/deleted``` after a ``` | ```?
About ```text/tabwriter```: yes! I didn't know it was a thing.

**Comment by josharian** [comment]
Yeah, maybe inserted/deleted annotations belong at the very end of the line.

**Comment by vekemans** [comment]
Ok, I've made some changes. What do you think?

File: cmd/semdiffstat/main.go [hunk 1]
```diff
@@ -49,22 +50,23 @@ func main() {
 		log.Fatalf("could not parse Go files %v and %v: %v\n", aName, bName, err)
 	}
 
+	w := new(tabwriter.Writer)
```
Comment (josharian) [inline cmd/semdiffstat/main.go]: You can do this in one line with `tabwriter.NewWriter`.

File: cmd/semdiffstat/main.go [hunk 2]
```diff
@@ -49,22 +50,23 @@ func main() {
...
+	w.Init(os.Stdout, 0, 8, 0, '\t', 0)
```
Comment (josharian) [inline cmd/semdiffstat/main.go]: Did you test this? I suspect you want something more like: `(os.Stdout, 0, 0, 0, ' ', 0)`

File: cmd/semdiffstat/main.go [hunk 3]
```diff
@@ -49,22 +50,23 @@ func main() {
...
 	for _, c := range changes {
-		// TODO: align | characters
 		// TODO: general UI improvements
 		if c.Inserted {
-			fmt.Printf("%v (inserted) %s\n", c.Name, pluses(c.InsLines))
+			fmt.Fprintf(w, "%v\t | %s (inserted)\n", c.Name, pluses(c.InsLines))
```
Comment (josharian) [inline cmd/semdiffstat/main.go]: I think you want a trailing `\t` at the end of each printed line.

**Review by josharian — CHANGES_REQUESTED** [review]
Thanks. text/tabwriter is a bit tricky; here are a few comments based on my experience with it.

I also now wonder whether we should just mimic git diff --stat entirely. Here's sample output:

```
 cmd/semdiffstat/main.go | 24 ++++++++++++++++--------
 doc.go                  |  1 +
 go.mod                  |  3 ++-
 go.sum                  | 12 ++++++++++++
```

Observe that they always write the total number of changes, and that those are also aligned. (I don't think text/tabwriter does sometimes-left sometimes-right alignment. Oh well. Two easy options there: decide it doesn't matter, or calculate max number of changes and pass it as a formatting adverb to Fprintf.)

But we can leave mimicking git diff --stat for another day / another PR, if you want to just get this one done.

File: cmd/semdiffstat/main.go (hunk 2 above)
Comment (josharian) [reply cmd/semdiffstat/main.go]: Thanks, GitHub, for a highly questionable line break decision here. The fifth parameter suggested is a space (`" "`).

File: cmd/semdiffstat/main.go (hunk 1 above)
Comment (vekemans) [reply cmd/semdiffstat/main.go]: Done

File: cmd/semdiffstat/main.go (hunk 2 above)
Comment (vekemans) [reply cmd/semdiffstat/main.go]: Done. I should I've read the docs better (it looked good on my screen though)

File: cmd/semdiffstat/main.go (hunk 3 above)
Comment (vekemans) [reply cmd/semdiffstat/main.go]: What do you mean? The output breaks if I add a trailing \t, looks fine otherwise.

**Comment by vekemans** [comment]
is this what you meant by mimicking ```git diff --stat```?

File: cmd/semdiffstat/main.go (hunk 3 above)
Comment (vekemans) [reply cmd/semdiffstat/main.go]: Oh yes, I see it in the docs...

**Comment by vekemans** [comment]
I don't think we can right-align the numbers column alone with tabwriter. It does look better though, I'll try to get it done tomorrow.

**Comment by vekemans** [comment]
The UI is now mimicking ```git diff --stat```, but I think the ```inserted```/```deleted``` text was useful.

File: cmd/semdiffstat/main.go [hunk 4]
```diff
@@ -79,3 +82,14 @@ func pluses(ins int) string {
 func minuses(del int) string {
 	return boldRed.Sprint(strings.Repeat("-", del))
 }
+func alignLines(changes []*semdiffstat.Change) (n int) {
```
Comment (josharian) [inline cmd/semdiffstat/main.go]: Call this maxCountLength?

File: cmd/semdiffstat/main.go [hunk 5]
```diff
@@ -79,3 +82,14 @@ func pluses(ins int) string {
...
+	var lens int
+	for _, c := range changes {
+		lens = c.InsLines + c.DelLines
+		lens = len(strconv.Itoa(lens))
```
Comment (josharian) [inline cmd/semdiffstat/main.go]: If we called the return value `max`, we could use `n` here. And no need to declare at the start of the function. So instead:

```
n := len(strconv.Itoa(c.InsLines+c.DelLines))
if n > max {
  max = n
}
```

File: cmd/semdiffstat/main.go [hunk 6]
```diff
@@ -49,22 +51,23 @@ func main() {
...
 		// TODO: general UI improvements
-		if c.Inserted {
-			fmt.Printf("%v (inserted) %s\n", c.Name, pluses(c.InsLines))
-			continue
-		}
-		if c.Deleted {
-			fmt.Printf("%v (deleted) %s\n", c.Name, minuses(c.DelLines))
-			continue
-		}
+		//if c.Inserted {
```
Comment (josharian) [inline cmd/semdiffstat/main.go]: Let's keep the annotations, but put them at the end. Something like:

```go
var annotation string
switch {
  case c.Inserted:
    annotation = " (inserted)"
  case c.Deleted:
    annotation = " (deleted)"
}
```

and then in the Fprintf, add a "%s" to the format string like `"%v\t | %*d\t %s%s%s\t\n",` and pass in annotation as an arg after minuses.

File: cmd/semdiffstat/main.go [hunk 7]
```diff
@@ -49,22 +51,23 @@ func main() {
...
+		//	fmt.Fprintf(w, "%v\t | %s (inserted)\n", c.Name, pluses(c.InsLines))
+		//	continue
+		//}
+		//if c.Deleted {
```
Comment (josharian) [inline cmd/semdiffstat/main.go]: Please delete dead code rather than commenting it out.

**Review by josharian — CHANGES_REQUESTED** [review]
Almost there, I think. Thanks for sticking with this.

**Comment by vekemans** [comment]
Done!

**Comment by josharian** [comment]
Thanks again.

**Pull request merged**